docs/proposals/hpa/cronfederatedhpa.md

```diff
@@ -0,0 +1,380 @@
+	Status CronFederatedHPAStatus `json:"status"`
+}
+
+type CronFederatedHPASpec struct {
+	// ScaleTargetRef points to the target resource to scale.
+	// It could be FederatedHPA, Deployment, StatefulSet, or other resource with scale sub-resource.
+	// +required
+	ScaleTargetRef autoscalingv2.CrossVersionObjectReference `json:"scaleTargetRef"`
+
+	// Rules is the list of CronFederatedHPARule,
+	// it define when and how to scale the related resources.
+	Rules []CronFederatedHPARule `json:"rules"`
+}
+
+type CronFederatedHPARule struct {
+	// RuleName is the name of the CronFederatedHPARule.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=32
+	// +required
+	RuleName string `json:"ruleName"`
+
+	// Schedule is the schedule of the CronFederatedHPARule.
+	// The format should follow https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#schedule-syntax
+	// +required
+	Schedule string `json:"schedule"`
+
+	// TargetReplicas is the target replicas to be scaled for deployment, statefulset, or
+	// other resources that have scale sub-resource.
+	// It should be set if ScaleTargetRef is not FederatedHPA.
+	// It shouldn't be set if ScaleTargetRef is FederatedHPA.
+	// +optional
+	TargetReplicas *int32 `json:"targetReplicas,omitempty"`
+
+	// TargetMinReplicas is the target MinReplicas to be set for FederatedHPA.
+	// If you not set this field, the value will not be updated.
+	// It shouldn't be set if ScaleTargetRef is not FederatedHPA.
+	// It should be set if ScaleTargetRef is FederatedHPA.
+	// +optional
+	TargetMinReplicas *int32 `json:"targetMinReplicas,omitempty"`
+
+	// TargetMaxReplicas is the target MaxReplicas to be set for FederatedHPA.
+	// If you not set this field, the value will not be updated.
+	// It shouldn't be set if ScaleTargetRef is not FederatedHPA.
+	// It should be set if ScaleTargetRef is FederatedHPA.
+	// +optional
+	TargetMaxReplicas *int32 `json:"targetMaxReplicas,omitempty"`
+
+	// Suspend is used to suspend the CronFederatedHPARule.
+	// The suspended rule will not be executed.
+	// Default is false
+	// +optional
+	Suspend *bool `json:"suspend,omitempty"`
+
+	// TimeZone for the giving rule, if not specified, the system time zone will be used.
+	// the system time zone comes from the running os of karmada-controller-mamanger component.
+	// Invalid TimeZone will be rejected when applying by karmada-webhook.
+	// see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones for the all timezones.
+	// +optional
+	TimeZone *string `json:"timeZone,omitempty"`
+}
+
+type CronFederatedHPAStatus struct {
+	// History is the execution history of the CronFederatedHPA rules.
+	RuleStates []RuleState `json:"ruleStates"`
+}
+
+type RuleState struct {
+	// RuleName is the name of the CronFederatedHPARule.
+	RuleName string `json:"ruleName"`
+	// NextExecutionTime is the next time to execute the CronFederatedHPARule.
+	NextExecutionTime *metav1.Time `json:"nextExecutionTime"`
```

Comment:
I suggest making `NextExecutionTime` an optional field, so that it won't be necessary once we introduce features like `RunOnce` in the future.